pkg/common/log/reopen.go

```diff
@@ -0,0 +1,91 @@
+package log
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+const (
+	fileFlags = os.O_APPEND | os.O_CREATE | os.O_WRONLY
+	fileMode  = 0640
+)
+
+var _ ReopenableWriteCloser = (*ReopenableFile)(nil)
+
+// Reopener inspired by https://github.com/client9/reopen
+type Reopener interface {
+	Reopen() error
+}
+
+type ReopenableWriteCloser interface {
+	Reopener
+	io.WriteCloser
+}
+
+type ReopenableFile struct {
+	name string
+	f    *os.File
+	mu   sync.Mutex
+}
+
+func NewReopenableFile(name string) (*ReopenableFile, error) {
+	file, err := os.OpenFile(name, fileFlags, fileMode)
+	if err != nil {
+		return nil, err
+	}
+	return &ReopenableFile{
+		name: name,
+		f:    file,
+	}, nil
+}
+
+func (r *ReopenableFile) Reopen() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	newFile, err := os.OpenFile(r.name, fileFlags, fileMode)
+	if err != nil {
+		r.f = nil
```

Comment:
I don't think we want to clear this on failure to open... shouldn't we continue trying to log to the existing file?

Reply:
Wonder if we shouldn't kick off a goroutine to log a warning that we couldn't rotate due to failure to open the file...

Reply:
> Wonder if we shouldn't kick off a goroutine to log a warning that we couldn't rotate due to failure to open the file...

Where would we log to if there is no log file - stderr?

Reply:
> I don't think we want to clear this on failure to open... shouldn't we continue trying to log to the existing file?
Yes. Seems obvious when you put it that way.

> Wonder if we shouldn't kick off a goroutine
I think we shouldn't. See my response in the other comment thread.

Reply:
What is your concern kicking off a goroutine that tries to log the failure to open? If logging is working, it is a short lived goroutine. If logging isn't working (and is blocking for whatever reason), this one goroutine will be the least of the server's problems.

Reply:
I'm not totally opposed to making the attempt to log to old file descriptor in a goroutine, but do we need to do that in a goroutine? I implemented that inline.

Reply:
The problem there is that the output won't be formatted correctly e.g. with a timestamp, json encoded, etc.